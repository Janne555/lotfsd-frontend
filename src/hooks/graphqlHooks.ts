import { useMutation } from '@apollo/react-hooks'
import { CharacterSheetUpdateMutation, CharacterSheetUpdateMutationVariables } from '../../__generated__/apolloTypes/CharacterSheetUpdateMutation'
import { CHARACTER_SHEET_UPDATE_MUTATION } from '../constants'
import { updateCharacterSheet } from '../services'

function useCharacterSheetUpdate(characterId?: string) {
  return useMutation<CharacterSheetUpdateMutation, CharacterSheetUpdateMutationVariables>(CHARACTER_SHEET_UPDATE_MUTATION, {
    update: updateCharacterSheet(characterId),
  })
}

export {
  useCharacterSheetUpdate
}