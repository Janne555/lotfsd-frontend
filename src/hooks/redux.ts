import { useSelector as useReduxSelector, TypedUseSelectorHook } from 'react-redux'
import { RootState } from '../Redux/store'

const useSelector: TypedUseSelectorHook<RootState> = useReduxSelector

export {
  useSelector
}