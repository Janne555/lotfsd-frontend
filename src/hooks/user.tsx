import React, { ReactNode } from 'react'
import { post } from '../services'
import { ENDPOINTS } from '../constants'
import { HTTPError } from 'ky'

type User = {
  username: string
  token?: string
}

type LoginContext = {
  user?: User
  login: (username: string, password: string) => void
  logout: () => void
  error?: string
}

const LoginContext = React.createContext<LoginContext | undefined>(undefined)

type ProviderProps = {
  children: ReactNode
  user?: User
}

function LoginProvider({ children }: ProviderProps): JSX.Element {
  const [user, setUser] = React.useState<User>()
  const [error, setError] = React.useState<string>()

  React.useEffect(() => {
    const username = window.localStorage.getItem("username")
    const token = window.localStorage.getItem("token")

    if (username && token) {
      setUser({
        username,
        token
      })
    }
  }, [])

  async function login(username: string, password: string) {
    setError(undefined)
    try {
      const response = await post(ENDPOINTS.LOGIN, { json: { username, password } })
      const token = await response.text()
      setUser({ username, token })
      window.localStorage.setItem("token", token)
      window.localStorage.setItem("username", username)
    } catch (error) {
      console.error(error)
      if (error instanceof HTTPError && error.response.status === 401) {
        setError("Wrong username or password")
      } else {
        setError("Login failed")
      }
    }
  }

  function logout() {
    setUser(undefined)
    window.localStorage.clear()
  }

  return (
    <LoginContext.Provider value={{
      login,
      logout,
      user,
      error
    }}>
      {children}
    </LoginContext.Provider>
  )
}

function useLogin() {
  const userContext = React.useContext(LoginContext)
  if (!userContext) {
    throw Error("Must be used within user context")
  }
  return userContext
}

function useLoginStatus(): LoginStatus {
  const login = useLogin()

  if (!login.user) {
    return 'logged-out'
  }

  if (!login.user.token) {
    return 'logging-in'
  }

  return 'logged-in'
}

function useToken(): string {
  const login = useLogin()
  if (!login.user?.token) {
    throw Error("Not logged in")
  }

  return login.user.token
}

function useUsername(): string {
  const login = useLogin()
  if (!login.user) {
    throw Error("not logged in")
  }

  return login.user.username
}

export {
  LoginContext,
  LoginProvider,
  useLoginStatus,
  useToken,
  useUsername,
  useLogin
}