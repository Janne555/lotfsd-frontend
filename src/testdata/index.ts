import { ItemsQuery } from "../../__generated__/apolloTypes/ItemsQuery"
import { MockedResponse } from "@apollo/react-testing"
import { CharacterListQuery } from "../../__generated__/apolloTypes/CharacterListQuery"
import { CHARACTER_LIST_QUERY, ITEMS_QUERY } from "../constants"

const ATTRIBUTES: Attributes = {
  charisma: 10,
  constitution: 11,
  dexterity: 12,
  intelligence: 13,
  strength: 14,
  wisdom: 15
}

const WALLET: Wallet = {
  copper: 10,
  gold: 20,
  silver: 30
}

const EQUIPMENT: EquipmentListItem[] = [
  {
    encumbrancePoints: 1,
    amount: 4,
    name: "thingy",
    itemId: "1",
    listItemId: "2",
    stackSize: 4
  },
  {
    encumbrancePoints: 4,
    amount: 4,
    name: "fooby",
    itemId: "3",
    listItemId: "4",
    stackSize: 4
  },
  {
    encumbrancePoints: 1,
    amount: 4,
    name: "baar",
    itemId: "5",
    listItemId: "6",
    stackSize: 4,
    encumbrance: 1
  }
]

type Result<T> = {
  data: T
}

const CHARACTER_LIST_MOCK: Result<CharacterListQuery> = {
  data: {
    characterSheets: [
      {
        __typename: "CharacterSheet",
        class: "fighter",
        id: "213123",
        name: "zoid"
      }
    ]
  }
}

const ITEM_INDEX_MOCK: Result<ItemsQuery> = {
  data: {
    items: [
      {
        __typename: "Item",
        attackBonus: null,
        baseArmorClass: null,
        damage: null,
        description: "a description",
        effects: null,
        encumbrance: null,
        encumbrancePoints: 1,
        id: "asd",
        name: "thingy",
        rangeLong: null,
        rangeMedium: null,
        rangeShort: null,
        stackSize: 1,
        type: "item"
      }
    ]
  }
}

const APOLLO_MOCKS: MockedResponse[] = [
  {
    request: {
      query: CHARACTER_LIST_QUERY
    },
    result: CHARACTER_LIST_MOCK
  },
  {
    request: {
      query: ITEMS_QUERY
    },
    result: ITEM_INDEX_MOCK
  }
]

export {
  ATTRIBUTES,
  WALLET,
  EQUIPMENT,
  APOLLO_MOCKS
}