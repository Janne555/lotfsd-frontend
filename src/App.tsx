import React from 'react'
import Main from './Main'
import { Switch, Route } from 'react-router-dom'
import { createApolloClient } from './services'
import { useLoginStatus, useToken } from './hooks'
import Login from './Components/Interface/Login'
import Register from './Components/Interface/Register'
import { ApolloProvider } from '@apollo/react-hooks'
import 'normalize.css'
import './index.css'

function App() {
  const loginStatus = useLoginStatus()
  const token = useToken()
  const client = React.useMemo(() => token ? createApolloClient(token) : undefined, [token])

  if (loginStatus === "logged-in" && client) {
    return (
      <ApolloProvider client={client}>
        <Main />
      </ApolloProvider>
    )
  } else {
    return (
      <Switch>
        <Route path="/register">
          <Register />
        </Route>
        <Route>
          <Login />
        </Route>
      </Switch>
    )
  }
}

export default App