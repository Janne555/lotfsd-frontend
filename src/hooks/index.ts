import { useDispatch, useSelector } from './redux'
import { useState, useLayoutEffect } from 'react'

function useScreenResizeEvent(condition: (screenWidth: number) => boolean) {
  const [result, setResult] = useState(condition(window.screen.width))
  useLayoutEffect(() => {
    function listener() {
      const now = condition(window.screen.width)
      if (now !== result) {
        setResult(now)
      }
    }
    window.addEventListener('resize', listener)
    return () => window.removeEventListener('resize', listener)
  })
  return result
}

export {
  useDispatch,
  useSelector,
  useScreenResizeEvent
}