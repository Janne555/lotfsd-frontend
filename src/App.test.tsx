import React from 'react'
import { JssProvider } from 'react-jss'
import { render, waitForDomChange } from '@testing-library/react'
import { MockedProvider } from '@apollo/react-testing'
import { MemoryRouter } from 'react-router-dom'
import App from './App'
import { LoginProvider } from './hooks'
import { APOLLO_MOCKS } from './testdata'

describe('<App />', () => {
  it('should match snapshot', async () => {
    const { asFragment } = render(
      <MemoryRouter>
        <LoginProvider>
          <MockedProvider mocks={APOLLO_MOCKS}>
            <JssProvider disableStylesGeneration>
              <App />
            </JssProvider>
          </MockedProvider>
        </LoginProvider>
      </MemoryRouter>
    )
    
    expect(asFragment()).toMatchSnapshot()
  });
});