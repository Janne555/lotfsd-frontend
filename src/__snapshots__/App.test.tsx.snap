// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<App /> should match snapshot 1`] = `
<DocumentFragment>
  <div>
    <div>
      <h1>
        Lamentations
        <br />
         of the
        <br />
        frontend developer
      </h1>
      <form>
        <div
          class="MuiFormControl-root MuiTextField-root"
        >
          <label
            class="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-outlined"
            data-shrink="false"
            for="username"
            id="username-label"
          >
            Username
          </label>
          <div
            class="MuiInputBase-root MuiOutlinedInput-root MuiInputBase-formControl"
          >
            <input
              aria-invalid="false"
              autocomplete="username"
              class="MuiInputBase-input MuiOutlinedInput-input"
              id="username"
              type="text"
              value=""
            />
            <fieldset
              aria-hidden="true"
              class="PrivateNotchedOutline-root-61 MuiOutlinedInput-notchedOutline"
              style="padding-left: 8px;"
            >
              <legend
                class="PrivateNotchedOutline-legend-62"
                style="width: 0.01px;"
              >
                <span>
                  ​
                </span>
              </legend>
            </fieldset>
          </div>
        </div>
        <div
          class="MuiFormControl-root MuiTextField-root"
        >
          <label
            class="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-outlined"
            data-shrink="false"
            for="password"
            id="password-label"
          >
            Password
          </label>
          <div
            class="MuiInputBase-root MuiOutlinedInput-root MuiInputBase-formControl"
          >
            <input
              aria-invalid="false"
              autocomplete="current-password"
              class="MuiInputBase-input MuiOutlinedInput-input"
              id="password"
              type="password"
              value=""
            />
            <fieldset
              aria-hidden="true"
              class="PrivateNotchedOutline-root-61 MuiOutlinedInput-notchedOutline"
              style="padding-left: 8px;"
            >
              <legend
                class="PrivateNotchedOutline-legend-62"
                style="width: 0.01px;"
              >
                <span>
                  ​
                </span>
              </legend>
            </fieldset>
          </div>
        </div>
        <button
          class="MuiButtonBase-root MuiButton-root MuiButton-outlined"
          tabindex="0"
          type="submit"
        >
          <span
            class="MuiButton-label"
          >
            Login
          </span>
          <span
            class="MuiTouchRipple-root"
          />
        </button>
        <button
          class="MuiButtonBase-root MuiButton-root MuiButton-text"
          tabindex="0"
          type="button"
        >
          <span
            class="MuiButton-label"
          >
            Register
          </span>
          <span
            class="MuiTouchRipple-root"
          />
        </button>
        <label
          class="MuiFormLabel-root Mui-error"
        />
      </form>
    </div>
  </div>
</DocumentFragment>
`;
